tests/simulator/py/fluid_init.py

```diff
@@ -0,0 +1,239 @@
+
+        sum_pop_density = self._sum_pop_patches(pop_densities, "density")
+        ions_density = self._get_contig_data_4_level0_ions_type(fluids, "ions_density")[
+            "density"
+        ]
+        np.testing.assert_allclose(
+            sum_pop_density, ions_density, rtol=tolerance,
+        )
+
+        sum_pop_flux = {
+            key: self._sum_pop_patches(pop_fluxs, key) for key in ["x", "y", "z"]
+        }
+        ions_bulkVelocity = self._get_contig_data_4_level0_ions_type(
+            fluids, "ions_bulkVelocity"
+        )
+        for key, val in ions_bulkVelocity.items():
+            np.testing.assert_allclose(
+                sum_pop_flux[key] / sum_pop_density, val, rtol=tolerance
+            )
+
+    def _checkFluidFromFluxFN(self, fluids, fn_name="flux"):
+        sim, model_dict = ph.globals.sim, ph.globals.sim.model.model_dict
+        pops = self._get_contig_popdata_4_level0_ions_type(
+            fluids, FluidInitValidation.pop_diag_file(fn_name)
+        )
+
+        def x0(pop, key, x):
+            return model_dict[pop]["v" + key](x) * model_dict[pop]["density"](x)
+
+        datasets, std_devs = self._get_std_devs_for(sim, pops, model_dict, x0)
+        dev_per_ppc = self._sort_std_devs_per_ppc(pops, std_devs, model_dict)
+        noises = self._get_noise_from(dev_per_ppc)
+        nppc = [x[1] for x in dev_per_ppc]
+        ratios = {key: noise[0] * np.sqrt(nppc[0]) for key, noise in noises.items()}
+        self.assertTrue(
+            all([diff < 0.09 for diff in self._get_diff(ratios, nppc, noises).values()])
+        )
+        for key, noise in noises.items():
+            self._save_std_fig(
+                nppc, noise, ratios[key], fn_name + key, test_threashold=0.007
+            )
+        self._save_dbg_fig(datasets, model_dict, fn_name)
+
+    def _checkFluidFromDensityFN(self, fluids, fn_name="density"):
+        sim, model_dict = ph.globals.sim, ph.globals.sim.model.model_dict
+        pops = self._get_contig_popdata_4_level0_ions_type(
+            fluids, FluidInitValidation.pop_diag_file(fn_name)
+        )
+
+        def x0(pop, key, x):
+            return model_dict[pop][key](x)
+
+        datasets, std_devs = self._get_std_devs_for(sim, pops, model_dict, x0)
+        dev_per_ppc = self._sort_std_devs_per_ppc(pops, std_devs, model_dict)
+        noises = self._get_noise_from(dev_per_ppc)
+        nppc = [x[1] for x in dev_per_ppc]
+        ratios = {key: noise[0] * np.sqrt(nppc[0]) for key, noise in noises.items()}
+        self.assertTrue(self._get_diff(ratios, nppc, noises)[fn_name] < 0.09)
+        self._save_std_fig(nppc, noises[fn_name], ratios[fn_name], fn_name)
+        self._save_dbg_fig(datasets, model_dict, fn_name)
+
+    def _get_contig_data_4_level0_ions_type(self, fluids, Type):
+        diags = Diagnostics.get(fluids, Type)
+        # can't be more than one, shouldn't be 0
+        assert len(diags) == 1
+        return self._aggregate_physical_patch_value(
```

Comment:
```
self._aggregate_physical_patch_value(sorted(diags[0].levels[0].patches, key=lambda x: x.min_coord("x"),), True,)
```

is too complicated to read

Reply:
better but what about:

```
patches = diags[0].levels[0].patches.values()
patches = sorted(patches, key=lambda x: x.min_coord("x"))
return _aggregate_physical_patch_value(patches, shared_patch_border=True)
```

Reply:
_aggregate_physical_patch_value has been moved to patch.py and changed to take a patch level